pkg/action/failover.go

```diff
@@ -0,0 +1,384 @@
+		logEvents := ac.printFunc(action, string(storkv1.ActionStatusFailed))
+		logEvents(msg, "err")
+		action.Status.Status = storkv1.ActionStatusFailed
+		action.Status.Reason = msg
+		ac.updateAction(action)
+		return
+	}
+
+	migrationNamespaces, err := utils.GetMergedNamespacesWithLabelSelector(migrationSchedule.Spec.Template.Spec.Namespaces, migrationSchedule.Spec.Template.Spec.NamespaceSelectors)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to get list of namespaces from migrationschedule %s/%s", migrationSchedule.Namespace, migrationSchedule.Name)
+		log.ActionLog(action).Errorf(msg)
+		ac.recorder.Event(action,
+			v1.EventTypeWarning,
+			string(storkv1.ActionStatusScheduled),
+			msg)
+		ac.updateAction(action)
+		return
+	}
+
+	if reactivate {
+		// in case of reactivation you have to activate the apps which were scaled down in the source cluster i.e. remote cluster
+		config, err = getClusterPairSchedulerConfig(migrationSchedule.Spec.Template.Spec.ClusterPair, migrationSchedule.Namespace)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to get the remote config from the clusterpair %s", migrationSchedule.Spec.Template.Spec.ClusterPair)
+			logEvents := ac.printFunc(action, string(storkv1.ActionStatusFailed))
+			logEvents(msg, "err")
+			action.Status.Status = storkv1.ActionStatusFailed
+			action.Status.Reason = msg
+			ac.updateAction(action)
+			return
+		}
+	} else {
+		config = ac.config
+	}
+
+	failoverSummaryList := make([]*storkv1.FailoverSummary, 0)
+	scaleupStatus := true
+	for _, ns := range namespaces {
+		// scale replicas only if namespace is a subset of namespaces being migrated
+		logEvents := ac.printFunc(action, "ScaleReplicas")
+		logEvents(fmt.Sprintf("Scaling up apps in cluster %s", config.Host), "out")
+		var failoverSummary *storkv1.FailoverSummary
+		if slices.Contains(migrationNamespaces, ns) {
+			err := resourceutils.ScaleReplicasReturningError(ns, true, false, ac.printFunc(action, "ScaleReplicas"), config)
+			if err != nil {
+				scaleupStatus = false
+				msg := fmt.Sprintf("scaling up apps in namespace %s failed: %v", ns, err)
+				log.ActionLog(action).Errorf(msg)
+				failoverSummary = &storkv1.FailoverSummary{Namespace: ns, Status: storkv1.ActionStatusFailed, Reason: msg}
+			}
+			msg := fmt.Sprintf("scaling up apps in namespace %s successful", ns)
+			failoverSummary = &storkv1.FailoverSummary{Namespace: ns, Status: storkv1.ActionStatusSuccessful, Reason: msg}
+		} else {
+			msg := fmt.Sprintf("skipping scaling up apps in namespace %s since it is not one of the namespaces being migrated by the MigrationSchedule %s/%s", ns, migrationSchedule.Namespace, migrationSchedule.Name)
+			failoverSummary = &storkv1.FailoverSummary{Namespace: ns, Status: storkv1.ActionStatusSuccessful, Reason: msg}
+		}
+		failoverSummaryList = append(failoverSummaryList, failoverSummary)
+	}
+
+	if reactivate {
+		// marking the stage's status as successful even if it hit error while activating apps up
```

Comment:
explain the rationale here.

Reply:
I had kind o followed what was written in the failback side. But on a second thought even I am wondering that it would make sense to atleast let the user know if reactivation failed, which namespaces were not reactivated. So that they can go ahead and try to activate those namespaces manually.
@diptiranjanpx what do you think about this ?

Reply:
The reactivate (essentially rollback) will only happen if something from previous stage has failed , so irrespective of the rollback stage's status, we have to mark the status as failed.

We can mark the reactivate stage's status `Failed` if something fails there. Anyway we will go to next stage irrespective of this stage succeeds or fails. 

We can raise the event saying which all reactivations failed but let's not put that in the reason field as it will have the info of original error.

Reply:
Also you can fill up the each namespace's status and reason in case of reactivation. That way at least we can have both the original failed reason and activation failure in the CR too.